### PR TITLE
101 color cube metal

### PR DESCRIPTION
**Pull request by loosche**
Implement a new mtl_renderer class that allows us to implement the 101_color_cube_metal experiment using the metal-cpp libraries from Apple.

### CODE REVIEW

File: projects/common/mtl_renderer.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    return success;
+}
+
+NS::Error* CreateBuffer(
+    MetalRenderer* pRenderer,
+    size_t         srcSize,
+    const void*    pSrcData,
+    MetalBuffer*   pBuffer)
+{
+    NS::AutoreleasePool* pPoolAllocator = NS::AutoreleasePool::alloc()->init();
```
Comment (chaoticbob) [inline projects/common/mtl_renderer.cpp:130]: Any idea why this is needed? It doesn't look like there are any other objects being created besides the buffer.

File: projects/common/mtl_renderer_utils.mm [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#include "mtl_renderer_utils.h"	
+
+#include <Cocoa/Cocoa.h>	
+#include <QuartzCore/QuartzCore.h>	
+
+void MetalSetNSWindowSwapchain(
```
Comment (chaoticbob) [inline projects/common/mtl_renderer_utils.mm:6]: nit: pointer vars missing `p` prefix.

File: projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    // *************************************************************************
+    // Compile shaders
+    // *************************************************************************
+    MetalShader vsShader;
+    MetalShader fsShader;
+    {
+        NS::Error*    pError   = nullptr;
+        MTL::Library* pLibrary = renderer->Device->newLibrary(
```
Comment (chaoticbob) [inline projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp:96]: Seems like a possible case for NS::SharedPtr?

**Review by chaoticbob — CHANGES_REQUESTED** [review]

File: projects/common/mtl_renderer.cpp (hunk 1 above)
Comment (loosche) [reply projects/common/mtl_renderer.cpp:130]: I'm not 100% sure any of the AutoreleasePools are needed? The example code has them so I assume that within their boundaries OSX won't delete any pointers? So the reason I have it is to make sure that OSX doesn't delete the pLocalBuffer before I can transfer it over to the NS::SharedPtr.

But those are all assumptions that may be invalid?

File: projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp (hunk 3 above)
Comment (loosche) [reply projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp:96]: I don't think we need it? I believe the AutoreleasePool will guarantee that nothing gets deleted within it, and we don't need the pLibrary pointer after the shaders are compiled. At least, I don't think we use it later (that would be a bug and require an NS::SharedPtr for sure).

File: projects/common/config.h [hunk 4]
```diff
@@ -16,6 +16,11 @@
 using Microsoft::WRL::ComPtr;
 #endif
 
+#if defined(__APPLE__)
+#include <float.h>
+#include <math.h>
```
Comment (chaoticbob) [inline projects/common/config.h:21]: Can we use `<cfloat>` and <cmath>` here?

File: projects/common/mtl_renderer.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        assert(false && "MTL::Device::newCommandQueue() failed");
+        return false;
+    }
+
+    return true;
+}
+
+bool InitSwapchain(
+    MetalRenderer*   pRenderer,
+    void*            cocoaWindow,
```
Comment (chaoticbob) [inline projects/common/mtl_renderer.cpp:51]: `pCocoaWindow`?

File: projects/common/mtl_renderer.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+#define GREX_DEFAULT_RTV_FORMAT MTL::PixelFormatBGRA8Unorm_sRGB
+#define GREX_DEFAULT_DSV_FORMAT MTL::PixelFormatDepth32Float
+
+struct MetalRenderer
+{
+    bool                                     DebugEnabled = false;
+    NS::SharedPtr<MTL::Device>               Device;
+    NS::SharedPtr<MTL::CommandQueue>         Queue;
+    CA::MetalLayer*                          Swapchain = nullptr;
```
Comment (chaoticbob) [inline projects/common/mtl_renderer.h:16]: `pSwapchain`?

File: projects/common/mtl_renderer.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    std::vector<NS::SharedPtr<MTL::Texture>> SwapchainDSVBuffers;
+    uint32_t                                 SwapchainBufferCount = 0;
+
+    MetalRenderer();
+    ~MetalRenderer();
+};
+
+bool InitMetal(MetalRenderer* pRenderer, bool enableDebug);
+bool InitSwapchain(MetalRenderer* pRenderer, void* cocoaWindow, uint32_t width, uint32_t height, uint32_t bufferCount = 2, MTL::PixelFormat dsvFormat = MTL::PixelFormatInvalid);
```
Comment (chaoticbob) [inline projects/common/mtl_renderer.h:25]: `pCocoaWindow`?

File: projects/common/mtl_renderer_utils.mm [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+void MetalSetNSWindowSwapchain(	
+   void* pCocoaWindow,	
+   void* pCAMetalLayer)	
+{	
+   NSWindow* pNSWindow = reinterpret_cast<NSWindow*>(pCocoaWindow);	
+   CAMetalLayer* pSwapchain = reinterpret_cast<CAMetalLayer*>(pCAMetalLayer);	
+
+   pNSWindow.contentView.layer = pSwapchain;
+   pNSWindow.contentView.wantsLayer = YES;
```
Comment (chaoticbob) [inline projects/common/mtl_renderer_utils.mm:14]: Does clang-format no like `.mm` files?

File: projects/common/window.cpp [hunk 9]
```diff
@@ -3,6 +3,11 @@
 #include <cassert>
 #include <fstream>
 
+#if defined(__APPLE__)
+#include <unordered_map>
```
Comment (chaoticbob) [inline projects/common/window.cpp:7]: Anything actually use from this include?

File: projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+		float3 position = vertexData.PositionOS;
+		output.PositionCS = Cam.MVP * float4(position, 1.0f);
+		output.Color = vertexData.Color;
+		return output;
+	}
+
+	float4 fragment fragmentMain( VSOutput in [[stage_in]] )
+	{
+		return float4(in.Color, 1.0);
+	}
```
Comment (chaoticbob) [inline projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp:62]: This shouldn't be indented.

File: projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    // *************************************************************************
+    // Main loop
+    // *************************************************************************
+    MTL::ClearColor clearColor(0.23f, 0.23f, 0.31f, 0);
+    uint32_t        frameIndex = 0;
+
+    while (window->PollEvents()) {
+        CA::MetalDrawable* pDrawable = renderer->Swapchain->nextDrawable();
+
+        // nextDrawable() will return nil if there are no free swapchain buffers to render to
```
Comment (chaoticbob) [inline projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp:174]: Does `nextDrawable()` wait at all?

**Review by chaoticbob — CHANGES_REQUESTED** [review]
Looks good just some minor formatting related stuff and some questions.

File: projects/common/window.cpp (hunk 9 above)
Comment (loosche) [reply projects/common/window.cpp:7]: I thought so, but apparently not. BALEETED!

File: projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp (hunk 11 above)
Comment (loosche) [reply projects/geometry/101_color_cube_metal/101_color_cube_metal.cpp:174]: It does, by default setAllowableTimeout() is true, which waits for 1 second. As we talked about I'm going to assert if it doesn't give back a drawable as after a second it probably means some error has occurred.

**Review by chaoticbob — APPROVED** [review]

**Pull request merged**